File: Arrays/P032.py
def optimal(nums):
    nums.sort(key=lambda x:x[0])
    merge=[]
    for interval in nums:
        if not merge or merge[-1][1]<interval[0]:
            merge.append(interval)
        else:
            merge[-1][0]=min(merge[-1][0],interval[0])
            merge[-1][1]=min(merge[-1][1],interval[1])
    return merge

intervals = [[1,3],[2,6],[8,10],[15,18]]
print("Optimal: ",optimal(intervals))

File: Striver_C++_Problems/Graphs/P032.cpp
// P032.cpp - Merge intervals

#include <bits/stdc++.h>
using namespace std;

vector<pair<int,int>> optimal(vector<pair<int,int>> nums) {
    sort(nums.begin(), nums.end(),
         [](const pair<int,int>& a, const pair<int,int>& b){
             return a.first < b.first;
         });

    vector<pair<int,int>> merge;
    for (auto interval : nums) {
        if (merge.empty() || merge.back().second < interval.first) {
            merge.push_back(interval);
        } else {
            merge.back().first  = min(merge.back().first,  interval.first);
            merge.back().second = min(merge.back().second, interval.second);
        }
    }
    return merge;
}

int main() {
    vector<pair<int,int>> intervals = {{1,3},{2,6},{8,10},{15,18}};
    auto res = optimal(intervals);
    cout << "Optimal: [";
    for (size_t i = 0; i < res.size(); i++) {
        cout << "[" << res[i].first << "," << res[i].second << "]";
        if (i + 1 < res.size()) cout << ", ";
    }
    cout << "]\n";
    return 0;
}
